meltdown.hs

```diff
@@ -21,18 +24,13 @@ type DeadSquares = [Square]
 ------------------ Instances so we can show things easily, sort, and check equality ------------------
 
 instance Show Square where
-	show (Square c r w h ch)		= show w ++ "x" ++ show h ++ " tile of character '" ++
-										[ch] ++ "' located at (" ++ show c ++ "," ++ show r ++ ")"
+	show (Square c r w h ch) = printf "%dx%d tile of character '%c' located at (%d, %d)" w h ch c r
 
 instance Eq Square where
-	(==) (Square col _ _ _ ch) (Square col2 _ _ _ ch2)
-									= (col == col2) && (ch == ch2)
+	(Square col _ _ _ ch) == (Square col2 _ _ _ ch2) = (col == col2) && (ch == ch2)
 
 instance Ord Square where
-	x `compare` y = if byRow == EQ then byChar else byRow
-		where
-			byRow = (startColumn x) `compare` (startColumn y)
-			byChar = (theChar x) `compare` (theChar y)
+        compare = comparing startColumn <> comparing theChar
```

Comment:
My changes to Ord are a mix of 2 improvements you may be interested in learning.  Firstly: the 

	if byRow == EQ then byChar else byRow

is a common haskell pattern: If a compare function's result is equal; fallback to another compare function.  If that is equal, fallback to another and so on...  This is captured in the Monoid typeclass for Ordering: EQ <> EQ <> LT <> GT = LT (used 2 fallback functions before finding something other than EQ). By extension, function that produce Orderings --- aka the 'compare' function are also Monoids!  With just that change your code becomes:

	x `compare` y = byRow x y <> byChar x y
		where		
			byRow x y = (startColumn x) `compare` (startColumn y)		
			byChar x y = (theChar x) `compare` (theChar y)


Secondly, the individual comparators (byRow, byChar) are applying the same function (startColumn in byRow) to each of the arguments and applying a function between them.  This is yet another common haskell abstraction: the 'on' function.  From the code above, you can convert:

	byRow x y = (startColumn x) `compare` (startColumn y)		

	byRow = compare `on` startColumn

And again, this is so common there is a function from Data.Ord:

	comparing = compare `on`

After both of these changes, you finally get:

	compare = comparing startColumn <> comparing theChar